### ISSUE TITLE
bounds_of_expr_in_scope() gives unexpected result for u32_sat()

### ISSUE THREAD

**Issue by steven-johnson** [halide/Halide#3271]
Given:

```
    Var x;
    Expr f32 = cast<float>(Load::make(Float(32), "buf", x, Buffer<>(), Parameter(), const_true()));
    Expr u32 = ConciseCasts::u32_sat(f32);
    Scope<Interval> scope;
    FuncValueBounds fb;
    Interval result_u32 = bounds_of_expr_in_scope(u32, scope, fb);
```

f32 is loaded from a buffer, and correctly has an expected bounds of [neg_inf, pos_inf]. 

u32 should be the result of a saturating cast to uint32, and thus I'd expect it to be [0, 0xffffffff]; however, it is actually [neg_inf, pos_inf].

signed-int32 is treated as 'never overflows' (thus a similar result there wouldn't be crazy), but I didn't think the same should be true for uint32.

**Comment by abadams** [comment]
The cause is that bounds_of_type in Bounds.cpp doesn't handle uint32. Not sure why this is, but it may just be historical (e.g. if we used int32s to represent the bounds at some point).

**Comment by steven-johnson** [comment]
Interesting. I'll try adding that support and see how things look.

**Comment by steven-johnson** [comment]
That looks like a good start, but I don't think it's enough: the problem here is that u32_sat() ends up generating IR something like

```
# u32 = u32_sat(f32)
let t1 = max(f32, 0.f)
let u32 = select(t1 >= 4294967296.f, uint32(4294967295), uint32(t1)))
```
but the cast node which is the false branch of the select (`uint32(t1)`) is only half-bounded [0, pos_inf], and the logic in the Cast node handler for Bounds.cpp refuses to propagate half-bounded intervals. (Is this deliberate? If so, it's unfortunate, since the 'parent' node would always make it fully bounded.)

**Comment by steven-johnson** [comment]
Ah, yes: I misread the output. It's definitely better; by fixing bounds_of_type() and nothing else, that expression becomes bounded, with an interval of 

```
min: select((buf[v0] < 4294967296.000000f), (uint32)0, (uint32)4294967295)
max:(uint32)4294967295
```

This seems much better, but that min looks dubious to me: it appears to be the result of injecting select()'s with more select()'s (see Bounds.cpp:757 or so). Wouldn't we be even better off sniffing for constants there and ending up with a min of (uint32)0?

**Comment by steven-johnson** [comment]
EDIT: nevermind, check_constant_bounds does indeed produce the forced-constant version.

**Comment by steven-johnson** [comment]
Yeah, my proposed fix is definitely inadequate; it ends up giving u32->i32 bogus intervals of [0, -1]. 

(For what it's worth, this is a real bug that is probably in need of fixing: I originally found it while adding `strict_float` to some code and was surprised to see `accessed in an unbounded way` errors... but discovered that the same code compiled as-is (without `strict_float`) but with just some clamps added could be induced to fail in the same way. Basically, doing u32->f32 conversions can end up being bounded or unbounded depending on how you structure the code, which is a recipe for unhappiness.)

At this point I doubt I have time to figure out and land a robust, proper fix before I leave for vacation, so I may see if I can create a workaround on my side, and leave this for when I return (or for someone else to ponder while I'm gone).
